c/dao.c

```diff
@@ -0,0 +1,290 @@
+#include "ckb_syscalls.h"
+#include "protocol_reader.h"
+
+#undef ns
+#define ns(x) FLATBUFFERS_WRAP_NAMESPACE(Ckb_Protocol, x)
+
+#define ERROR_UNKNOWN -1
+#define ERROR_WRONG_NUMBER_OF_ARGUMENTS -2
+#define ERROR_SYSCALL -4
+#define ERROR_BUFFER_NOT_ENOUGH -10
+#define ERROR_ENCODING -11
+#define ERROR_OVERFLOW -12
+#define ERROR_INVALID_WITHDRAW_BLOCK -13
+#define ERROR_INCORRECT_CAPACITY -14
+
+#define HASH_SIZE 32
+#define DAO_SIZE 32
+#define HEADER_SIZE 4096
+/* 32 KB */
+#define WITNESS_SIZE 32768
+
+#define LOCK_PERIOD_BLOCKS 10
+#define MATURITY_BLOCKS 5
+
+#define MAX(a, b) (((a) < (b)) ? (b) : (a))
+
+static int extract_bytes(ns(Bytes_table_t) bytes, unsigned char *buffer,
+                         volatile size_t *s) {
+  flatbuffers_uint8_vec_t seq = ns(Bytes_seq(bytes));
+  size_t len = flatbuffers_uint8_vec_len(seq);
+
+  if (len > *s) {
+    return ERROR_BUFFER_NOT_ENOUGH;
+  }
+
+  for (size_t i = 0; i < len; i++) {
+    buffer[i] = flatbuffers_uint8_vec_at(seq, i);
+  }
+  *s = len;
+
+  return CKB_SUCCESS;
+}
+
+/*
+ * Fetch withdraw header hash from the 3rd(offset by 1) argument
+ * of witness table. Kept as a separate function so witness buffer
+ * can be cleaned as soon as it is not needed.
+ */
+static int extract_withdraw_header_index(size_t input_index, size_t *index) {
+  int ret;
+  volatile uint64_t len = 0;
+  unsigned char witness[WITNESS_SIZE];
+
+  len = WITNESS_SIZE;
+  ret = ckb_load_witness(witness, &len, 0, input_index, CKB_SOURCE_INPUT);
```

Comment:
Check that `len <= WITNESS_SIZE`

Reply:
ℹ️**Review Information**

Each withdraw consumes a former deposited cell as an input. The corresponding witnesses contain two arguments:

- The first is intended to be consumed by the lock script to prove the user owns the cell.
- The second one is a u64 encoded in little endian, it is the index into the transaction deps, which references an anchor header. We assume that the cell is withdraw from that anchor header. Transaction submitter should use the most recent block as the anchor header. The interest between anchor header to the block really commit the withdraw transaction is a implicit cost of the withdraw.